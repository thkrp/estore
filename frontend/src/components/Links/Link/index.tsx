import React, { AriaRole, FC } from 'react';
import { Link as ReactLink } from 'react-router-dom';
import { sanitizeUrl } from '@braintree/sanitize-url';

type Props = {
    to: string;
    children?: React.ReactNode;
    className?: string;
    title?: string;
    role?: AriaRole;
};
/**
 * The "Link" component from react-router-dom with a sanitized url.
 */
const Link: FC<Props> = ({ to = '', children, className, title, role }) => {
    return (
        <ReactLink to={sanitizeUrl(to)} className={className} title={title} role={role}>
            {children}
        </ReactLink>
    );
};

export default Link;
