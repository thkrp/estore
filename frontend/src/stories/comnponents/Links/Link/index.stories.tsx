import { Meta, StoryObj } from '@storybook/react';
import { withRouter } from 'storybook-addon-react-router-v6';
import { Link as LinkComponent } from '../../../../components';

const meta = {
    title: 'Components/Links/Link',
    component: LinkComponent,
    decorators: [withRouter],
    parameters: {
        layout: 'centered'
    },
    tags: ['autodocs'],
    argTypes: {
        to: {
            description: 'href value'
        },
        children: {
            description: 'React.ReactNode'
        },
        className: {
            description: 'class names'
        }
    }
} satisfies Meta<typeof LinkComponent>;

export default meta;
type Story = StoryObj<typeof meta>;
export const Link: Story = {
    args: {
        to: '/',
        children: 'link text'
    }
};
