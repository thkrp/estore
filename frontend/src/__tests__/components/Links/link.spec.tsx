/* eslint-disable no-restricted-imports*/
import React from 'react';
import '@testing-library/jest-dom';
import { cleanup, render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import * as sanitizer from '@braintree/sanitize-url';
import Link from '../../../components/Links/Link';

describe('Link', () => {
    afterEach(cleanup);

    it('sanitizeUrl has been called', () => {
        jest.spyOn(sanitizer, 'sanitizeUrl');
        const { container } = render(<Link to="/" />, { wrapper: MemoryRouter });
        expect(sanitizer.sanitizeUrl).toHaveBeenCalled();
        expect(container).toMatchSnapshot();
    });

    it('should be able to pass a class', () => {
        render(<Link to="/" className="sanitized-link active" />, { wrapper: MemoryRouter });
        const element = screen.getByRole('link');
        const classes = element.className.split(' ');
        expect(classes.includes('sanitized-link')).toBe(true);
        expect(classes.includes('active')).toBe(true);
    });

    it('should be able to pass a title', () => {
        render(<Link to="/" title="sanitized-link" />, { wrapper: MemoryRouter });
        const element = screen.getByRole('link');
        const { title } = element;
        expect(title).toBe('sanitized-link');
    });

    it('should be able to pass a children', () => {
        render(
            <Link to="/">
                <div>LINK</div>
            </Link>,
            { wrapper: MemoryRouter }
        );
        expect(screen.getByText(/LINK/)).toBeInTheDocument();
    });
});
