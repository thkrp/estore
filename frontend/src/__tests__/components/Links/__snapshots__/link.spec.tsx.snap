// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link sanitizeUrl has been called 1`] = `
<div>
  <a
    href="/"
  />
</div>
`;
